src/NHibernate/Hql/Ast/ANTLR/QueryTranslatorImpl.cs

```diff
@@ -191,11 +213,9 @@ private void ErrorIfSelect()
 				throw new QueryExecutionRequestException("Not supported for select queries:", _queryIdentifier);
 			}
 		}
-
-		public NHibernate.Loader.Loader Loader
-		{
-			get { return _queryLoader; }
-		}
+		
+		// 6.0 TODO : change type to ILoader
+		public Loader.Loader Loader => _queryLoader as Loader.Loader ?? throw new NotSupportedException("Loader is not of supported type.");
```

Comment:
This property is used by the multi-query batcher (so, by futures too). This implementation means custom query loaders not deriving from the Loader class will not be supported when using the query batching feature. (Either explicitly through the use of a query batch or implicitly through the use of futures.)

Can you add some test about this?

Reply:
For supporting them, it will likely be needed to add an `ILoader QueryLoader` property to this class, obsolete the `Loader` one and use the new property "everywhere" the older one was used.

Of course adding that QueryLoader property to the IQueryTranslator interface would be a breaking change, so if in need of adding it there, a new transitional interface would be needed.

Reply:
You are right, this implementation causes some troubles with Future-Queries. 

Here is a possible solution, where the underlying QueryLoader of a decorated QueryLoader can be exposed and used:
```suggestion

		// 6.0 TODO : change type to ILoader
		public Loader.Loader Loader => _queryLoader.GetUnderlyingQueryLoader() as Loader.Loader ?? throw new NotSupportedException("Loader is not of supported type.");

	/// <summary>
	/// Defines a query loader.
	/// </summary>
	public partial interface IQueryLoader: ILoader
	{	
		// 6.0 TODO : Remove when Loader.Loader properties are replaced by ILoader properties.
		/// <summary>
		/// Returns the underlying query loader.
		///
		/// E.g. When a query loader gets decorated, it is possible to return the decoratee here. 
		/// </summary>
		/// <returns></returns>
		IQueryLoader GetUnderlyingQueryLoader();
	}
	
	public partial class QueryLoader : BasicLoader, IQueryLoader
	{
		// 6.0 TODO : Remove when Loader.Loader properties are replaced by ILoader properties.
		/// <summary>
		/// Returns the underlying query loader which is the current implementation.
		/// </summary>
		/// <returns></returns>
		public virtual IQueryLoader GetUnderlyingQueryLoader()
		{
			return this;
		}
	}
	
	internal sealed class CustomQueryLoader: IQueryLoader
	{
		private readonly IQueryLoader _queryLoader;

		public CustomQueryLoader(IQueryLoader queryLoader)
		{
			_queryLoader = queryLoader;
		}

		public IQueryLoader GetUnderlyingQueryLoader()
		{
			return _queryLoader;
		}
	}
	
	Test:
		[Test]
		public void FutureQueryTest()
		{
			var customers = _session.CreateQuery("select c from Customer c")
			                        .Future<Customer>()
			                        .ToList();
			
			Assert.AreEqual(1, customers.Count);
		}
```

Reply:
That does not sound to me as complete fix, and it seem more complex than what I propose for implementors. (But allow for less changes in NHibernate code.) I'll try to spare some time to push a commit with what I think could be a better solution.